### PR TITLE
Add error handler to publish error messages via MQTT

### PR DESCRIPTION
**Pull request by ch-evergen**
## What?

- Add ErrorHandler and ErrorMessage classes, to take error message data, serialise as JSON and publish to MQTT
- Provide the ErrorHandler to classes which will generate runtime errors (not used for startup)
- Use separate MQTT Reader and Writer classes, each with an independent MQTT client

In a separate PR I will add:
- end-to-end testing to demonstrate error publishing
- including site and serial details in error topic

## Why?

We need a way to get information on runtime errors back to the cloud from the IPC.

## Testing/Proof

Manual testing using mosquitto_sub/pub shows error messages being published. Test coverage is still at 100%. More E2E testing coming soon.

The following screenshots demonstrate error messages being published and received. The terminal windows shown are, top to bottom:
- a terminal where I manually execute `mosquitto_pub` commands to emulate commands being transmitted to the MQTT broker from the cloud
- a docker-compose process running mosquitto and pymodbus
- a python process running remote-commands-handler `main.py`
- a `mosquitto_sub` process which receives messages sent to the `error/#` topic

First, sending a well-formed command. This passes through `main.py` without logging and results in output from `pymodbus` when the command is received at the modbus server.

Screen Shot 2023-06-22 at 10 31 49 am

Second, sending a command with an invalid JSON structure - there is no `value` key. This results in an `InvalidMessage` error being logged in `main.py` and published to MQTT:

Screen Shot 2023-06-22 at 10 32 18 am

Third, sending a command which is valid JSON but references a command name that is not configured on this system. This results in an `UnknownCommand` error, also logged in `main.py` and published to MQTT:

Screen Shot 2023-06-22 at 10 32 33 am

### CODE REVIEW

File: main.py
Comment (serroba) [inline main.py:145]: do you need to pass the entire configuration here? wouldn't you only need a string or simple type instead?

File: main.py
Comment (serroba) [reply main.py:148]: you could split the responsibilities though. They don't have to depend on the same thing. Have separate abstractions to subscribe and publish and each having it's own instantiation process.
What do you need from the `mqtt_client`? I think you only care about the connection. We could, hypothetically, be using different connections as well. To subscribe from mosquitto and to write to something else.
I wonder if you can instead just keep them separated from each other. They both would depend on the same library (paho?), but besides that, they share little in common

File: main.py
Comment (ch-evergen) [reply main.py:145]: I prefer to pass the entire config in all cases, generally in the interest of minimising the code in main.py (which is not covered by unit tests or coverage checks). For error handler there's some conditional logic where the relevant config settings depend on whether the error topic is set or not, and I'd rather do that checking inside error handler itself than in main.py.

File: main.py
Comment (ch-evergen) [reply main.py:148]: hmm so perhaps instead of a single `MQTTClient` module, we would have an `MQTTReader` and `MQTTWriter` ? seems like it might be a bit overkill - the writer would need to manage its own connection rather than just using the existing one. If we actually want to enable using two different MQTT brokers for commands and errors, then we'd need a different set of MQTT config settings as well. I think unless there's a compelling reason to allow two separate connections, I'd rather keep it simple and stick with the slightly clunky method of passing mqtt client to error handler after initialising.

File: main.py
Comment (serroba) [reply main.py:145]: yeah, I see your point here 👍

File: main.py
Comment (serroba) [reply main.py:148]: I believe I am a bit biased here by what telegraf does, but I think is worth discussing this further.
In general, I feel that objects should have _all_ their dependencies on instantiation. If you can't do that, is often, if not always, a problem with the design. Now, on saying that, I am not saying that the only way to have a better design is by doing what I am suggesting, but I'd like to avoid the partial state on the constructor as a goal. The how we achieve that, remains open, but I will try to explain my perspective.

image

The current class is responsible to act, not as a generic MQTT client, but as a very opinionated, dedicated abstraction to take JSON documents. It has a main `run` method already.

We could make the class objective more generic (renaming the `run` method), but I would argue that would just be reducing cohesion of the class. The class is now doing more disparate functions. I tend to feel that SRP is a nice principle here. On saying that, we could also decide to rename the class/file, to make it more explicit that the objective is to act as the `MQTT ingestor`  instead of a generic abstraction around MQTT.


Telegraf has 4 main abstraction and everything is essentially an implementation of one of those abstraction. You have input data, output data, transformers and serialisers. The inputs and outputs remain separated from each other. I like that approach, as allows them to evolve independently from each other. 

For the longer objectives we have, at some point, we will have to think about multiple device protocol support, not only Modbus. Sure, I am here talking about something else, this abstraction is about MQTT, not Modbus. But probably is a means to explain that we may need to think about having possible more than way to back-propagate error messages, not only via MQTT. I would like to have an abstraction that's responsible to handle errors that's not necessarily coupled to way we receive input data.

Comment (serroba) [reply main.py:148]: Also, worth mentioning that is probably incorrect to expose a public method to allow the client to determine where to subscribe. To be clear here, I am talking about a parallel task, not the one you are actively working on, but I would recommend to make private the `subscribe_topics` method
Upon instantiation, once you receive the config file, you can subscribe this abstraction to the correct topic.
No need to let the client orchestrate that.

Comment (ch-evergen) [reply main.py]: OK, still some tidying up to do, but I now have 3 separate MQTT classes:
- base class `MQTTClient` which only connects
- `MQTTReader` which subscribes to topics and runs a loop
- `MQTTWriter` which connects when needed to publish an error message

It works ok and there's not too much duplication of code, and it means the error handler and mqtt reader can be initialised without circular dependencies in `main.py`. Note this is only because the writer module doesn't take an error handler, on the assumption that if it encounters any errors, we're by definition not going to be able to propagate them via MQTT. If we did want to still pass in an error handler and use other non-mqtt behaviour to publish errors, there would still be a circular dependency problem.

There are some awkward bits, like the fact that to mock the internal MQTT client inside error handler, we have to pass the mock down through two layers (first to error_handler init, then to MQTTWriter init). I probably could refine this to make the 'error reporting mechanism' a bit more abstract, but I'm not sure how much benefit there is in further abstraction. Also there's still some coverage to add and docs to update. Let me know what you think :)

File: app/mqtt_reader.py
Comment (serroba) [inline app/mqtt_reader.py:22]: Whats the need to use inheritance over composition here?
I don't think we need to think of the reader and writer of siblings.

I would instead recommend to think of them as 2 distinct actions, that happen to depend on the same dependency upon instantiation. But that's encapsulated

File: app/mqtt_reader.py
Comment (serroba) [inline app/mqtt_reader.py:37]: as I think I said in a different place. I would rather make this method private and call it upon instantiation. I think you have everything when you 

Rather than doing this in main:

```
mqtt_client = setup_mqtt_client(configuration)

    mqtt_client.subscribe_topics([configuration.mqtt_settings.command_topic])
```
we should just do instantiation of the object and upon instantiation, we know where to connect.

File: app/mqtt_writer.py
Comment (serroba) [inline app/mqtt_writer.py:15]: same thing as above. I like the idea to keep them separated. I would strongly recommend to use composition over inheritance though

File: app/mqtt_writer.py
Comment (serroba) [inline app/mqtt_writer.py:27]: similar to what I commented above about the reader, the topic, is somehting we should be able to know upon instantiation.
However, because we would like to allow to send error messages to different MQTT suffixes, what I think we should let the client indicate, is what's the suffix. So, rather than receiving the entire `topic`, we should rename this as `suffix: string`

Comment (serroba) [reply app/mqtt_writer.py:27]: Also, can we add a type to the `payload` property?

File: app/mqtt_reader.py
Comment (ch-evergen) [reply app/mqtt_reader.py]: Good call. I removed this method altogether as it was confusingly named - it was just populating an internal attribute rather than actually doing the subscribe operation.

File: app/mqtt_reader.py
Comment (ch-evergen) [reply app/mqtt_reader.py]: I've refactored to have MQTTReader and MQTTWriter instantiate an MQTTClient class which they use to connect (I stored it in `self.connector` rather than `self.client` because the multiple levels of "client" were confusing). But I actually don't think it's worth it - the only useful thing left in the class is a single `connect` method and I think repeating that within reader and writer classes is an acceptable level of duplication.

Comment (serroba) [reply app/mqtt_reader.py]: yes, that's fine. We can inject the same instance. In this case, paho remains an implementation detail of each object. There is not exposure of the MQTT interface to the clients of the reader or error handler.
We can swap or change things in the future and clients shouldn't care about such changes.

Comment (ch-evergen) [reply app/mqtt_reader.py]: OK that's done, MQTTClient is removed and both reader and writer have their own connect method.

And it may be asking for trouble, but I reworked ErrorHandler to use a separate class when it's using MQTT to transmit errors. The previous implementation required several optional args to `init` because the MQTT error handling may or may not be activated. I think this way is clearer, but happy to adjust if inheritance is not appropriate here either :)

File: app/mqtt_writer.py
Comment (ch-evergen) [reply app/mqtt_writer.py:31]: What would `type` be used for? Is it similar to `category` or additional?

I can see a bit of confusion here ... I have two `publish` methods:
- one in `ErrorHandler` which is opinionated about the payload structure
- and one here in `MQTTWriter` which is trying to be quite generic (but still calls `ErrorMessage.write` to do the serialisation).
 
My thinking was, if `MQTTWriter` is used in future to publish messages to other non-error-related topics, it would need to accept the entire topic string rather than append to the prefix configured for errors - so I allowed the full topic to be specified here. But maybe this is premature and it's ok for `MQTTWriter` to assume that it's always publishing error messages? Or should it be completely unopinionated about topic/payload (in which case ErrorMessage should serialise the payload before passing it in)?

File: app/error_handler.py
Comment (serroba) [inline app/error_handler.py:21]: Is this needed?

File: main.py
Comment (serroba) [inline main.py:109]: What do we need a factory method? Why not using the constructor directly instead?

File: app/error_handler.py
Comment (serroba) [reply app/error_handler.py:21]: BTW there is a name for this
https://lostechies.com/chadmyers/2009/07/14/the-usual-result-of-poor-man-s-dependency-injection/
It's often not recommended.

The term though has two meanings. One is not having a container and do all the instantiation manually. The other is partially injected classes. I'm talking about the latter.

File: app/mqtt_writer.py
Comment (serroba) [reply app/mqtt_writer.py:31]: I think it's a valid question. I'm not strongly opinionated here and I see your point now.

The goal of this project though is not to be a bidirectional tool..l the goal is to send commands to the edge. Based on that, I feel the mqtt writer doesn't have to be generic.

I would even consider to avoid that abstraction and just encapsulate that inside the error handler. That's not needed though, having separation of concerns is also good. I'll let you to decide there. Having that in mind, I agree that the underlying layer if you keep it, should be generic. Acting more like a driver. The wrapper layer, the error handler class is the one that would be in charge to have deeper control and expose a dedicated interface

Comment (ch-evergen) [reply app/mqtt_writer.py]: ok thanks - I've kept MQTTWriter and moved the serialisation back to ErrorHandler, making this a completely generic driver for publishing to any MQTT topic - so I've left the topic as fully determined by the caller. I think it's worth keeping to encapsulate specifying the `qos` and handling the callback.

File: main.py
Comment (ch-evergen) [reply main.py:109]: The challenge with initialising the ErrorHandler is the variable number of required config settings: either you don't want to publish error messages (in which case `pub_errors` is false and you don't need to provide any other args to init), or you do (in which case you need to init with all the mqtt config settings and the paho client object).

The way I have it currently is that if you are not publishing messages, you use the base `ErrorHandler` class; if you are, you use a subclass `MQTTErrorHandler` which requires all those things in its init method. The factory method determines which class of object you get back, and both init methods require the correct number of args for their mode of operation. Happy to hear suggestions if that's not the best way :)

Comment (serroba) [reply main.py:109]: Let's move with this.
If you want to merge the pr as is I can approve it after you update the pr title.

Otherwise, we are missing a demonstration of this working.

It's probably a good time for you to get familiar with the pymodbus server online modification.
We discussed previously you can initiate it in REPL mode and change its behaviour or you can do on the headless start by calling the http endpoint. The latter is good for end to end automation. The former is good for manual t testing.

Can you please ensure to provide evidence of this working as expected?

Comment (ch-evergen) [reply main.py:109]: I'm not sure that this is ready to merge yet - I think recent changes have introduced an issue, probably related to having the reader and writer classes connecting to the broker independently. After I publish an error message, the reader seems to lose its topic subscription. Still looking into it.

Comment (ch-evergen) [reply main.py:110]: OK that's sorted - the issue was that I was creating one `mqtt_client` and passing it to both reader and writer, which meant the callbacks overwrote each other and messed things up. It's back to working as expected now with two separate client objects.

File: app/error_handler.py
Comment (serroba) [inline app/error_handler.py:25]: This double nested if could be avoided if we always inject the mqtt client.

My preference remains on avoiding partial state but I won't block this pr. It's mostly where I think we should be, so I'm ok proceeding with this

Comment (serroba) [reply app/error_handler.py:25]: BTW one reason why I have this preference is because partial state is always evaluated on runtime rather than bootstrap time. By enforcing all dependencies to be instantiated as part of the constructor, you can do static analysis that shows errors. The app doesn't start if there is an error here. However, with partial state, the app does starts and is only on runtime when you can the accidental "null pointer exceptions"

File: app/mqtt_reader.py
Comment (serroba) [inline app/mqtt_reader.py:34]: Any reason to not have this as private?

File: app/mqtt_reader.py
Comment (serroba) [inline app/mqtt_reader.py:88]: Shouldn't we be throwing an exception here ?

File: main.py
Comment (serroba) [inline main.py:73]: I wonder if it is a good idea to encapsulate all this configuration behind a dedicated class instead of having this in main. What do you think?

File: main.py
Comment (serroba) [reply main.py:110]: I guess that says we were missing some end to end test then. Happy to add it as a follow up but we should let the pipeline be the one that shows those errors if could happen

**Review by serroba — APPROVED** [review]
Nice work @ch-evergen 
I have a few questions and comments but all those can be addressed latter on. 

It would be nice if you attach also an screenshot with the errors working in mosquitto

File: main.py
Comment (ch-evergen) [reply main.py:110]: yes good point - I'll modify the e2e test in #39 to ensure that we re-test receiving commands after publishing an error

File: app/error_handler.py
Comment (ch-evergen) [reply app/error_handler.py:25]: I see the point, but do you think it's OK to always require the MQTT client in the constructor, even if the site is configured not to publish error messages? I guess it doesn't affect the way we actually init ErrorHandler, but in general it seems to me that a class shouldn't require a dependency if it knows at bootstrap time (at least after reading config) that it's not going to use it.

File: app/mqtt_reader.py
Comment (ch-evergen) [reply app/mqtt_reader.py:88]: This is just the callback after connect is completed - if it had failed we would already have raised an `OSError` exception in the main `connect` method

File: app/error_handler.py
Comment (serroba) [reply app/error_handler.py:25]: What's the problem of doing that though?
The object exists regardless. It's indeed the same object after all.
It's just a pointer pointing to a real memory allocation rather than pointing nowhere.

On the other hand, the class becomes simpler. As a developer you don't need to worry to check if the dependencies exist or not. This often leads to less accidental runtime errors

Comment (serroba) [reply app/error_handler.py:25]: But I have approved your pr. I don't feel we must address this right now. Feel free to move on with the pr as is.

We may revisit this in the future but not a blocker now

File: app/mqtt_reader.py
Comment (ch-evergen) [reply app/mqtt_reader.py:34]: Hmmm. Making it private would indicate that it shouldn't be modified, but there might be a valid reason to be able to see the value. I'll make it private and we can add an accessor method later if needed

File: app/error_handler.py
Comment (ch-evergen) [reply app/error_handler.py:25]: Yes, thanks for that. Just trying to understand the reasoning as it doesn't feel intuitive to me. I'll attach those screenshots then merge this one, but might come back to this in #39

File: main.py
Comment (ch-evergen) [reply main.py:73]: The code that defines the command-line args is only ever going to be applicable to main.py, so there's no reusability value in having it in a separate class - but it does make it testable and would include it in coverage checks. Same for the code that combines the command-line args with the config settings. So it sounds like a hassle but probably a good idea :)

Comment (serroba) [reply main.py:73]: Yeah I get it. No strong feelings either way

**Pull request merged**
